src/python/pants/backend/docker/parser.py

```diff
@@ -0,0 +1,98 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os.path
+import re
+from dataclasses import InitVar, dataclass, field
+from typing import Generator
+
+from dockerfile import Command, parse_string
+
+from pants.backend.docker.target_types import DockerImageSources
+from pants.engine.fs import DigestContents, GlobMatchErrorBehavior, PathGlobs
+from pants.engine.rules import Get, collect_rules, rule
+
+_pex_target_regexp = re.compile(
+    r"""
+    (?P<path>(?:\w\.?)+) / (?P<name>(?:\w\.?)+) \.pex$
+    """,
+    re.VERBOSE,
+)
+
+
+@dataclass(frozen=True)
+class DockerfileParseRequest:
+    sources: DockerImageSources
+
+
+@dataclass(frozen=True)
+class DockerfileInfo:
+    putative_target_addresses: tuple[str, ...] = ()
+
+
+@dataclass
+class DockerfileParser:
+    dockerfile: InitVar[str]
+    commands: tuple[Command, ...] = field(init=False)
+
+    def __post_init__(self, dockerfile: str):
+        self.commands = parse_string(dockerfile)
+
+    def get_all(self, command_name: str) -> Generator[Command, None, None]:
+        for command in self.commands:
+            if command.cmd.upper() == command_name:
+                yield command
+
+    def translate_to_address(self, value: str) -> str | None:
+        # Translate something that resembles a packaged pex binary to its corresponding target
+        # address. E.g. src.python.tool/bin.pex => src/python/tool:bin
+        pex = re.match(_pex_target_regexp, value)
```

Comment:
Is there a better way to resolve what target produced a certain output?

Reply:
I don't think so. This is a little bit fragile though - if someone changes the address->path mapping this will break, and they won't know. So tests will be key!  Testing this function separately, and aggressively, will be important.

Reply:
Will do. Didn't spend time testing it until I was sure it was going to be used, in it's current form :)